demos/device_defender_for_aws/metrics_collector/freertos_plus_tcp/metrics_collector.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * FreeRTOS V202011.00
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://aws.amazon.com/freertos
+ * https://www.FreeRTOS.org
+ */
+
+/**
+ * @file metrics_collector.c
+ *
+ * @brief Functions used by the defender demo to collect metrics on the
+ * device's open ports and sockets. FreeRTOS+TCP tcp_netstat utility
+ * is used to collect this metrics.
+ */
+
+/* Standard includes. */
+#include <stdio.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "FreeRTOS_IP.h"
+
+/* FreeRTOS+TCP tcp_netstat utility include. */
+#include "tcp_netstat.h"
+
+/* Demo config. */
+#include "defender_demo_config.h"
+
+/* Interface include. */
+#include "metrics_collector.h"
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t GetNetworkStats( NetworkStats_t * pOutNetworkStats )
+{
+    MetricsCollectorStatus_t status = MetricsCollectorSuccess;
```

Comment:
It is almost always preferable to initialize return status variables to failure, e.g.:

```c
MetricsCollectorStatus_t status = MetricsCollectorCollectionFailed;
```

The success value should only be set when the code has passed the success path.  This also often makes your code more readable.  For example, this function would only have 1 `if` instead of 2 as in this pseudocode:

```
status = FAIL

if( vGetMetrics() == success ) {
    logdebug("got metrics")
    fill in stats
    status = SUCCESS
}
return status